### ISSUE TITLE
Support new DateOnly and TimeOnly structs in .NET 6

### ISSUE THREAD

**Issue by mattjohnsonpint** [JamesNK/Newtonsoft.Json#2521]
.NET 6 (starting with preview 4) will support two new primitives: `System.DateOnly` and `System.TimeOnly`.
Refer to: https://github.com/dotnet/runtime/issues/49036

These are similar to `System.DateTime`, but split into subcomponents.  They should serialize/deserialize in their ISO 8601 formats, such as `"2021-12-31"` and `"23:59:59.9999999"`, respectively.

It would be great if these were supported out of the box by Json.NET.  However, that would require targeting `net6.0`, and presently Json.NET is built for `netstandard` targets only.  Alternatively, a separate library could be built to add custom `JsonConverter` objects for these types, though it seems a pitfall that users won't know they will need that bit of glue.

Unfortunately, the types do not implement `[Serializable]` (See [comment](https://github.com/dotnet/runtime/issues/49036#issuecomment-789453258)).  So I believe they will need to have support added directly.

Any thoughts on this would be appreciated.  Thanks.

**Comment by mattjohnsonpint** [comment]
Converters would be like so:

```csharp
public class DateOnlyJsonConverter : JsonConverter<DateOnly>
{
    private const string DateFormat = "yyyy-MM-dd";

    public override DateOnly ReadJson(JsonReader reader, Type objectType, DateOnly existingValue, bool hasExistingValue, JsonSerializer serializer)
    {
        return DateOnly.ParseExact((string)reader.Value, DateFormat, CultureInfo.InvariantCulture);
    }

    public override void WriteJson(JsonWriter writer, DateOnly value, JsonSerializer serializer)
    {
        writer.WriteValue(value.ToString(DateFormat, CultureInfo.InvariantCulture));
    }
}

public class TimeOnlyJsonConverter : JsonConverter<TimeOnly>
{
    private const string TimeFormat = "HH:mm:ss.FFFFFFF";

    public override TimeOnly ReadJson(JsonReader reader, Type objectType, TimeOnly existingValue, bool hasExistingValue, JsonSerializer serializer)
    {
        return TimeOnly.ParseExact((string)reader.Value, TimeFormat, CultureInfo.InvariantCulture);
    }

    public override void WriteJson(JsonWriter writer, TimeOnly value, JsonSerializer serializer)
    {
        writer.WriteValue(value.ToString(TimeFormat, CultureInfo.InvariantCulture));
    }
}
```

**Comment by JamesNK** [comment]
Closer to .NET 6 release I think Newtonsoft.Json should:
* Add a net6.0 target
* Built-in support for serializing and deserializing DateOnly and TimeOnly
* Consider built-in support for these types in JValue.

**Comment by simonziegler** [comment]
@JamesNK, I just bumped into this issue, saving `DateOnly` properties to CosmosDB. Is there a resolution on the horizon given that .NET 6 is released?

**Comment by spreedated** [comment]
I took me an hour to figure out why my objects aren't compared correctly, through I have a custom comparer.
Problem is, the `DateOnly `struct from my json file renders `01.01.0001 ` which was the root of all the evil; switching to `DateTime` fixed the problem.

**Comment by craigomatic** [comment]
> @JamesNK, I just bumped into this issue, saving `DateOnly` properties to CosmosDB. Is there a resolution on the horizon given that .NET 6 is released?

It's clunky but you can decorate your properties with an attribute to the appropriate converter, ie what @mattjohnsonpint posted above. I had the same issue with ```TimeOnly``` + CosmosDB and this got things sorted:

```csharp
[JsonConverter(typeof(TimeOnlyJsonConverter))] 
public TimeOnly StartTime { get; set; }
```

**Comment by karl-sjogren** [comment]
@JamesNK Would you accept a PR for this?

Just implementing your first two points should be fairly easy (of course with proper tests and such).

* Add a net6.0 target
* Built-in support for serializing and deserializing DateOnly and TimeOnly

I don't mind spending some time on this if there is a chance to get it merged. Seeing how many open PRs there are however I feel like it might be a waste of time.

**Comment by JamesNK** [comment]
Yes, I’d accept a PR. I’ve been waiting to see how System.Text.Json does it so merging will probably wait until they have something.

**Comment by KenBonny** [comment]
I've run into the problem in an F# project I'm doing. So here is the F# equivalent

```fsharp
type DateOnlyJsonConverter() =
    inherit JsonConverter<DateOnly>()
    let format = "yyyy-MM-dd"

    override this.WriteJson(writer: JsonWriter, value: DateOnly, _: JsonSerializer): unit =
        writer.WriteValue(value.ToString(format, CultureInfo.InvariantCulture))

    override this.ReadJson(reader, _, _, _, _) =
        DateOnly.ParseExact(reader.Value :?> string, format, CultureInfo.InvariantCulture)
        
type TimeOnlyJsonConverter() =
    inherit JsonConverter<TimeOnly>()
    let format = "HH:mm:ss.FFFFFFF"

    override this.WriteJson(writer: JsonWriter, value: TimeOnly, _: JsonSerializer): unit =
        writer.WriteValue(value.ToString(format, CultureInfo.InvariantCulture))
    override this.ReadJson(reader, _, _, _, _) =
        TimeOnly.ParseExact(reader.Value :?> string, format, CultureInfo.InvariantCulture)
```

How to use:

```
type Person = {
  Name: string
  BirthDate: DateOnly
}

let json = JsonConvert.SerializeObject(person, Formatting.None, DateOnlyJsonConverter())
(* result:
  { "name": "Ken", "birthDate": "1987-05-20" }
*)
```

**Comment by bennycoomans** [comment]
> Yes, I’d accept a PR. I’ve been waiting to see how System.Text.Json does it so merging will probably wait until they have something.

@JamesNK: Maybe you've already seen it, but it has been fixed in System.Text.Json in PR https://github.com/dotnet/runtime/pull/69160.

**Comment by voroninp** [comment]
@JamesNK was this already released?

**Comment by sungam3r** [comment]
https://www.nuget.org/packages/Newtonsoft.Json/13.0.2-beta1

**Comment by angelaki** [comment]
Will this version ever make it stable? Just having an issue with serializing DateOnly using 13.0.1

**Comment by ekjuanrejon** [comment]
getting the following with cosmos db
`Newtonsoft.Json.JsonSerializationException: 'Error converting value 31/07/2023 00:00:00 to type 'System.Nullable`1[System.DateOnly]'. `

**Comment by vukasinpetrovic** [comment]
Is there a way to specify format for TimeOnly with latest version? I would like it to only show HH:mm, without seconds.

**Comment by angelaki** [comment]
If this is not supported out of the box (and I don't think so), I guess you could just write your own converter for this type?

**Comment by vukasinpetrovic** [comment]
@angelaki Yes, that's a solution I'm thinking to go for but I just wanted to confirm before I proceed with that.

**Comment by Sasino97** [comment]
It's 2025 and it's still not supported out of the box...
```
Newtonsoft.Json.JsonSerializationException: 'Error converting value "04:00" to type 'System.TimeOnly'. Path 'Schedules[0].Time', line 23, position 27.'
```

**Comment by h3x4d3c1m4l** [comment]
> It's 2025 and it's still not supported out of the box...
> 
> ```
> Newtonsoft.Json.JsonSerializationException: 'Error converting value "04:00" to type 'System.TimeOnly'. Path 'Schedules[0].Time', line 23, position 27.'
> ```

Fork it, fix the issue then create a PR :)

**Comment by DougSchmidt-Leeward** [comment]
@Sasino97 @h3x4d3c1m4l I've bumped into the same issue.

Looking at the implementation added in #2679 it seems that both `DateOnly` and `TimeOnly` serializers were added with support for only one specific format each.

- `DateOnly` is `value = DateOnly.ParseExact(s, "yyyy'-'MM'-'dd", CultureInfo.InvariantCulture);`
- `TimeOnly` is `value = TimeOnly.ParseExact(s, "HH':'mm':'ss.FFFFFFF", CultureInfo.InvariantCulture);`

Those format strings are not configurable.

I've used the sample snipper from @mattjohnsonpint in this thread and configured the list of acceptable format strings. That worked enough for me.
